### ISSUE TITLE
[REQ] [Java] Support of JakartaEE 9

### ISSUE THREAD

**Issue by RomainPruvostMHH** [OpenAPITools/openapi-generator#13124]
### Context

Since Eclipse Foundation is in charge of JakartaEE specifications, it's time to switch to JakartaEE 9 for openapi generator.
The first step was to switch to JakartaEE 8 (see the issue #10504). It's done now.

The next step is to upgrade to JakartaEE 9 which means to upgrade to new versions of jakarta's dependencies.
But this upgrade is a bit more difficult than the previous because the main goal of JakartaEE 9 is that the javax.* packages to jakarta.* packages have been renamed by Eclipse.
I think that it will be necessary to create a new property (named jakartaeeVersion=9 for example) to allow the user to generate the source code with the actual versions of jakartaEE 8 or the last versions of jakartaEE 9.

## Describe the solution you'd like

We need to upgrade the following dependencies in the Maven/Gradle/Scala mustache template files. These template files are located in openapi-generator/modules/openapi-generator/src/main/resources.

jakarta.annotation:jakarta.annotation-api:1.3.5 => 2.0.0
jakarta.validation:jakarta.validation-api:2.0.2 => 3.0.0
jakarta.ws.rs:jakarta.ws.rs-api:2.1.6 => 3.0.0
jakarta.json.bind:jakarta.json.bind-api:1.0.2 => 2.0.0
jakarta.json:jakarta.json-api:1.1.6 => 2.0.0
jakarta.xml.bind:jakarta.xml.bind-api:2.3.3 => 3.0.0
jakarta.el:jakarta.el-api:3.0.3 => 4.0.0
jakarta.servlet:jakarta.servlet-api:4.0.4 => 5.0.0
jakarta.activation:jakarta.activation-api:1.2.2 => 2.0.0

The upgrade of jakarta versions will force us to replace all javax.* import by jakarta.* in the mustache templates.

Then, we need a new property named jakartaeeVersion to keep the generation source code as is (with the package javax.* ) or to use the new jakartaEE 9 classes in the generated source code.
By default, if the property didn't set, we consider that we are still on jakartaEE 8 (javax.*).

Beware, some samples are dependents of Spring Boot. For them, it will be necessary to upgrade the version of spring boot parent to 3 (release in September 2022). It is not part of the scope of this issue.

## How to validate the request 

We need to create an unit test with the property jakartaeeVersion. The versions of depedencies should be equal to the versions above.
And of course, we need to create a new java sample with the new property.

## Additional context

Versions of JakartaEE 9 : https://projects.eclipse.org/releases/jakarta-ee-9. Or a map here https://jakarta.ee/release/9/
Explanations of JakartaEE 9 release plan : https://eclipse-ee4j.github.io/jakartaee-platform/jakartaee9/JakartaEE9ReleasePlan

**Comment by bmarwell** [comment]
This has been a request for two years now:

https://github.com/OpenAPITools/openapi-generator/issues/6881

**Comment by estigma88** [comment]
Spring Boot 3 release is moving from `javax.` to `jakarta.` and this is a blocker to upgrade, anything we can help with?

**Comment by madorb** [comment]
This is a big problem now that Spring Boot 3.x / Spring Framework 6.x are released and only support Java 17 + JakartaEE 9. OpenApi generator no longer will work with any up-to-date spring apps.

**Comment by PascalSchumacher** [comment]
Wasn't this feature already implemented in: https://github.com/OpenAPITools/openapi-generator/pull/12407 ?

**Comment by madorb** [comment]
i suppose it's not clear, from the ticket - but i believe server side support has been implemented, but client generation still does not support emitting JakartaEE 9 annotations (Generated, bean validation, etc.)

**Comment by arham-saeed** [comment]
This is a huge blocker for migration to spring boot 3.

**Comment by varminas** [comment]
This is a blocker for us as well. As an option would be the usage of spring generator, but it generates Java classes which is not the aim.

**Comment by SwapnilKhante** [comment]
Is there a work around here?

**Comment by RomainPro76** [comment]
As a workaround, you can do the following: 

```xml
            <plugin>
                <groupId>org.openapitools</groupId>
                <artifactId>openapi-generator-maven-plugin</artifactId>
                <version>6.2.1</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                        <configuration>
                            <inputSpec>
                                ${project.basedir}/src/main/resources/api.yml
                            </inputSpec>
                            <generatorName>spring</generatorName>
                            <apiPackage>com.api</apiPackage>
                            <modelPackage>com.api.model</modelPackage>
                            <generateApiTests>false</generateApiTests>
                            <generateModelDocumentation>false</generateModelDocumentation>
                            <generateModelTests>false</generateModelTests>
                            <generateSupportingFiles>false</generateSupportingFiles>
                            <configOptions>
                                <interfaceOnly>true</interfaceOnly>
                                <java8>true</java8>
                                <dateLibrary>java8</dateLibrary>
                                <performBeanValidation>true</performBeanValidation>
                                <useBeanValidation>true</useBeanValidation>
                                <serializationLibrary>jackson</serializationLibrary>
                                <delegatePattern>false</delegatePattern>
                                <skipDefaultInterface>true</skipDefaultInterface>
                            </configOptions>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
            <!-- com.google.code.maven-replacer-plugin - Workaround : waiting for openapi-generator-maven-plugin supports jakartaEE 9. -->
            <plugin>
                <groupId>com.google.code.maven-replacer-plugin</groupId>
                <artifactId>replacer</artifactId>
                <version>1.5.3</version>
                <executions>
                    <execution>
                        <id>replace-import-javax-by-jakarta</id>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>replace</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <basedir>
                        ${project.build.directory}/generated-sources/openapi/src/main/java/com/api
                    </basedir>
                    <includes>
                        <include>**/*.java</include>
                    </includes>
                    <replacements>
                        <replacement>
                            <token>import javax</token>
                            <value>import jakarta</value>
                        </replacement>
                    </replacements>
                </configuration>
            </plugin>
```

The plugin com.google.code.maven-replacer-plugin:replacer search "import javax" strings in the generated files and replace them with "import jakarta" strings.

**Comment by madorb** [comment]
anyone happen to know of a similar plugin for gradle, google isn't giving me much luck

**Comment by yudanaz** [comment]
I can confirm the "spring" generator solution, but couldn't make it work for the "java" generator (we use `library= 'webclient'` in our project) using 6.3.0-SNAPSHOT. I implemented a simple string replacement (`javax\\.` for `jakarta.`) in our Gradle build as a workaround. It does the job, but it's far from pretty.  
This issue needs high priority due to the Spring Boot 3 release!

**Comment by dsyer** [comment]
For me `library=webclient` is broken even with Spring Boot 2.7. Don't you get a generated uncompilable `ApiUtils` with references to `HttpServletRequest`?

**Comment by yudanaz** [comment]
That's strange, we use `generatorName = 'java'` with `library = 'webclient'` successfully. No `ApiUtil` generated (I only now that class from the spring generator). We were using Spring Boot 2.7.5 (now 3.0.1) and still are using open-api generator version 6.2.1.

**Comment by dsyer** [comment]
Another update: it works for me with just a client and Spring Boot 3. Here's the Maven config:

```xml
			<plugin>
				<groupId>org.openapitools</groupId>
				<artifactId>openapi-generator-maven-plugin</artifactId>
				<version>6.3.0-SNAPSHOT</version>
				<executions>
					<execution>
						<id>openapi-client</id>
						<phase>generate-sources</phase>
						<goals>
							<goal>generate</goal>
						</goals>
						<configuration>
							<generatorName>java</generatorName>
							<apiPackage>com.example.demo.client</apiPackage>
							<modelPackage>com.example.demo.client</modelPackage>
							<invokerPackage>com.example.demo.client</invokerPackage>
							<configOptions>
								<useJakartaEe>true</useJakartaEe>
								<library>webclient</library>
							</configOptions>
						</configuration>
					</execution>
				</executions>
				<configuration>
					<inputSpec>${project.basedir}/openapi.yaml</inputSpec>
					<configOptions>
						<performBeanValidation>true</performBeanValidation>
						<hideGenerationTimestamp>true</hideGenerationTimestamp>
						<useBeanValidation>true</useBeanValidation>
						<sourceFolder>src/main/java</sourceFolder>
						<configPackage>com.example.demo.config</configPackage>
					</configOptions>
				</configuration>
			</plugin>
```

**Comment by boardtc** [comment]
Thanks to you guys for trying to figure out workarounds for this serious unexpected blocker on moving to SpringBoot 3.
Do I need to use the snapshot version of the plugin? I am trying with 6.2.1 without messing with our build setup. 
Trying
`              <configOptions>
                <useJakartaEe>true</useJakartaEe>
                <library>webclient</library>
              </configOptions>`

I am getting:
java.lang.RuntimeException: Unknown library: spring-mvc

**Comment by dsyer** [comment]
I'm not sure what the "html" generator does with a "library" (it's not documented). Did it ever work? Anyway the examples above are all for the "spring" and "java" generators.

**Comment by boardtc** [comment]
Thanks for letting me know. We have always generated java even with that set to HTML! I tried replacing it with both spring and java and both gave the same error with the above settings :-(

**Comment by dsyer** [comment]
My sample is here if you want to crib the pom.xml config: https://github.com/dsyer/openapi-demo. If you share your minimal sample I might be able to see what you need.

**Comment by boardtc** [comment]
Thank you!!

I see you use 6.3.0-SNAPSHOT, so that definitely needed then over 6.2.1? What did you add in your settings.xml to allow that to download?

I tried your config, I was not using the outer <generaorName> for spring and I moved to use the resttemplate library.  My generated file does not import javax! But it does use @javax.annotation.Generated, is that wrong?

The build subsequently failed as [ERROR] Failed to execute goal org.apache.maven.plugins:maven-resources-plugin:3.3.0:resources (default-resources) on project ABC: filtering ...\src\main\resources\v1\gradle\wrapper\gradle-wrapper.jar to ...\target\classes\v1\gradle\wrapper\gradle-wrapper.jar failed with MalformedInputException: Input length = 1

Not sure yet if this related to the plugin.

**Comment by dsyer** [comment]
There's nothing in my settings.xml (AFAIK) it's all in the pom. If you read the comments above you will find that snapshots are only needed for the "java" generator.

If you are getting `javax` instead of `jakarta` namespaces then you aren't doing the same as my sample.

The thing with the gradle wrapper is weird (why is it in your classpath resources?), but looks completely unrelated to OpenAPI code generation.

**Comment by boardtc** [comment]
Thanks for the support!
Adding the <plugin repository> section from your code does not work for me as it does not bypass our nexus which does not have the snapshot. So I am still testing with 6.21.

We did not have a separate configuration outside of the 2 execution sections so I restyled it like yours. I did not have hideGenerationTimestamp or configPackage, which I added. We also have the additional:
`	        <generateSupportingFiles>true</generateSupportingFiles>

	        <dateLibrary>java8</dateLibrary> 

	        <interfaceOnly>true</interfaceOnly>

	        <skipOverwrite>false</skipOverwrite>

	        <useTags>true</useTags>` 
	        
But I get the same error...

**Comment by madorb** [comment]
The [Gradle wrapper](https://docs.gradle.org/current/userguide/gradle_wrapper.html) is not generated by a maven plugin. Is your project a maven or gradle project? It seems like maven, and if so, you almost certainly want to update your `.openapi-generator-ignore` to cause it to not generate either the maven or gradle files. 

Also then share the _entirety_ of your `<plugin>` configuration as dave has above, sharing little snippets doesn't help anyone help you.

**Comment by boardtc** [comment]
We do not use gradle at all, it's a java project

**Comment by dsyer** [comment]
> Is there anything in these settings that pulls gradle into the generation?

Sort of. You are using the `openapi-generator-maven-plugin` and by default it generates a full project with Maven and Gradle build configurations. You should be able to ignore them though. Your mistake will be something unrelated and unshown in the snippet above. A simple reproducer project in github would probably clear things up. It looks like maybe you are copying everything that the plugin generates into another location and that's probably not what you needed.

**Comment by boardtc** [comment]
Thanks for the reply. The way you say "I am using the openapi-generator-maven-plugin" suggests there may be another one?

This pom snippet worked with 5.0.0 previously and did not generate Gradle, so it looks like that was added in v6 to be done by default? That said 2 weeks ago [@madorb was looking for a simple plugin for gradle](https://github.com/OpenAPITools/openapi-generator/issues/13124#issuecomment-1369916992) even though this does one does gradle?

How do others stop Gradle from being generated by default? 

I need to figure out why the 6.3.0-SNANSHOT API is unrecognisable from the 5.0.0 generated one.

**Comment by dsyer** [comment]
> How do others stop Gradle from being generated by default?

I suspect they don't bother mostly. It hasn't been an issue for me - I just ignore some of the generated files.

**Comment by boardtc** [comment]
Fair enough. I don't care about Gradle being generated if it was not causing me an issue.

**Comment by boardtc** [comment]
With v5.0.0 I had only 1 generated java file in /v1/resources, with this version there is a whole project copied in there as [shown above](https://github.com/OpenAPITools/openapi-generator/issues/13124#issuecomment-1387187021). I can't see any change in my pom file that is doing this, does anyone know what has changed between 5.0.0 & 6.3.0 and does one now just have to populate a .openapi-generator-ignore file?

**Comment by boardtc** [comment]
As mentioned above I was using spring-mvc library with c5. I have found an [announcement ticket](https://github.com/OpenAPITools/openapi-generator/issues/11553) which says: 
![image](https://user-images.githubusercontent.com/214808/213755405-93e3982c-d18a-441f-846c-f8b6fa53b778.png)
As suggested I tried spring-boot, but it I get the error:
```
[ERROR] 
java.lang.RuntimeException: Unknown library: spring-boot
Available libraries:
  jersey1
  jersey2
  jersey3
  feign
  okhttp-gson
  retrofit2
  resttemplate
  webclient
  resteasy
  vertx
  google-api-client
  rest-assured
  native
  microprofile
  apache-httpclient
```
which should I use in place of spring-mvc?

**Comment by dsyer** [comment]
I would have said, probably you need to use the "spring" generator instead of the "java" generator, but that was in the pom snippet you already posted. If you don't want to post a complete sample it's going to be hard to piece it all together.

**Comment by boardtc** [comment]
Thanks for engaging!! I posted my [complete snippet](https://github.com/OpenAPITools/openapi-generator/issues/13124#issuecomment-1387451058) above when you asked...did I miss something?

I changed mine around (above) to match your demo so I now use spring instead of java in execution 1 and java instead of spring in execution 2 like your demo....

**Comment by dsyer** [comment]
> I am wondering if others do the opposite of what I am doing and generate straight to target and then only copy the java API back to resources?

You will see that pattern in the sample I posted. The default location for generated source files is in `target`. You can then pick the ones you need using the resource plugin (or any other mechanism that suits you).

An SSCCE is by definition *not* your full-scale "reasonably sized application". It is a sample that reproduces your problem. It can be a chore to create such a sample, but it makes everyone much happier in the long run. I suggest that this discussion is way off topic for the issue now, and you should find another forum (Stack Overflow maybe?) to continue.

**Comment by boardtc** [comment]
> The [Gradle wrapper](https://docs.gradle.org/current/userguide/gradle_wrapper.html) is not generated by a maven plugin. Is your project a maven or gradle project? It seems like maven, and if so, you almost certainly want to update your `.openapi-generator-ignore` to cause it to not generate either the maven or gradle files.

@matorb I have learned that the ignore file is to prevent files from being overwritten by the generator rather than being generated in the first place. There appears to be no way to control what's generated,

> I suggest that this discussion is way off topic for the issue now, and you should find another forum (Stack Overflow maybe?) to continue.
@dsyer ongoing issues with the plugin but I thank you for your patience and won't post here again

**Comment by wing328** [comment]
the release note only highlights some changes but not all.

please refer to https://github.com/OpenAPITools/openapi-generator/pulls?q=is%3Apr+is%3Amerged+milestone%3A6.3.0 for a full list of changes.

i'm still revising the release note.

**Comment by boardtc** [comment]
@wing328 thanks for letting me know.
With a maven clean install I get lots of compilation errors like:
/target/src/java/main/org/openapitools/client/Pair.java:[16,18] package javax.annotation does not exist

**Comment by wing328** [comment]
can you share more details on how you generate the code? e.g. spec, generator name, options, etc

**Comment by craig-day** [comment]
I can confirm that with this config and 6.3.0, I'm getting jakarta annotations

```groovy
openApiGenerate {
    generatorName = "java"
    inputSpec = "$buildDir/openapi.yaml"
    outputDir = projectDir.path
    packageName = "foo"
    apiPackage = "foo"
    modelPackage = "foo"
    ignoreFileOverride = "$projectDir/.openapi-generator-ignore"
    configOptions = [
            apiPackage: "foo",
            modelPackage: "foo",
            asyncNative: "true",
            documentationProvider: "none",
            library: "resteasy",
            microprofileRestClientVersion: "3.0",
            useJakartaEe: "true"
    ]
}
```

**Comment by boardtc** [comment]
> can you share more details on how you generate the code? e.g. spec, generator name, options, etc

Thanks for reply. My pom settings are [above](https://github.com/OpenAPITools/openapi-generator/issues/13124#issuecomment-1387451058)

**Comment by dsyer** [comment]
Can I suggest we close this issue and invite people to raise separate requests about specific combinations that do not currently work? Or maybe we could document which combinations actually do work a bit better somewhere. E.g. a table in the form

|Generator | Library | Flags | Supported |
|------|------|-----|--|
|spring       |            | useSpringBoot3=true | X|
|java          | resttemplate | useJakartaEe=true | X|
|java          | webclient | useJakartaEe=true | X|
|java          | resteasy | useJakartaEe=true | X|
|java          | jersey1    | useJakartaEe=true | ? |
|java          | jersey2    | useJakartaEe=true | ? |
|java          | jersey3    | useJakartaEe=true | ? |
|java          | feign    | useJakartaEe=true | ? |
|java          | okhttp-gson    | useJakartaEe=true | ? |
|java          | retrofit2    | useJakartaEe=true | ? |
|java          | vertx    | useJakartaEe=true | ? |
|java          | google-api-client    | useJakartaEe=true | ? |
|java          | rest-assured    | useJakartaEe=true | ? |
|java          | native    | useJakartaEe=true | ? |
|java          | microprofile    | useJakartaEe=true | ? |
|java          | apache-httpclient    | useJakartaEe=true | ? |

**Comment by bmarwell** [comment]
> Can I suggest we close this issue and invite people to raise separate requests about specific combinations that do not currently work? Or maybe we could document which combinations actually do work a bit better somewhere. E.g. a table in the form

Like this one from 2020? 😢 Already mentioned it a few times.
Which entry in this table would that be? It is not bound to any library!

https://github.com/OpenAPITools/openapi-generator/issues/6881

**Comment by dsyer** [comment]
Library is an optional configuration setting. I'd be happy to add other generators if you tell me you have tested it and it works.

**Comment by bmarwell** [comment]
That would be `jaxrs-spec` in your table. it should be good for reasteasy, jersey and cxf as well as they all implement jaxrs. Btw, jaxrs is now called "restfulWs" starting with 3.0 (jakarta namespace). So, implement one, get three :) Example is in the issue I mentioned earlier.

My apps will always only use the API (dependency scope=provided in maven) and the implementation only in the runtime. So yes, it is already heavily used. Some IBMers working on Open Liberty did comment in the issue I mentioned, too.

**Comment by dsyer** [comment]
I didn't see any actual solutions proposed in that other issue, but I'm happy to take your word for it if you say it works. I just don't know what, precisely, you are saying works already. I don't see any references to a generator called "restfulWs", and there *is* a "useJakartaEe" flag for "jaxrs-spec" but does it work, and is it needed? Has anyone actually tested the other "jaxrs-*" generators?

**Comment by boardtc** [comment]
I have been looking at this again and using 6.4.0, trying to get unblocked, when building after the OpenAPI Generator finishes the maven-compiler-plugin detects changes and recompiles, fine except it then reverts to javax.
The property mentioned above jakartaeeVersion has not yet been added. I am setting useJakartaEe to true. Why is javax still being pulled in via [{{javaxPackage}}](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/BeanValidationException.mustache) rather than Jakarta?

org/openapitools/client/BeanValidationException.java:[5,24] package javax.validation does not exist
`package org.openapitools.client;

import java.util.Set;

import javax.validation.ConstraintViolation;
import javax.validation.ValidationException;

public class BeanValidationException extends ValidationException {
`

**Comment by dsyer** [comment]
It would really be better to post this as a question on stackoverflow and/or provide a full working sample. Which of the matrix combinations in my table above are you using?

**Comment by boardtc** [comment]
It's spring boot so java & useJakartaEe with webClient.

I'm unsure why here is not suitable since people are trying to get workarounds for functionality which has not been implemented yet in the plugin.

**Comment by dsyer** [comment]
The way I read it, the config in that link has two executions with `[generatorName,library,options]` equal to `[java,,useSpringBoot3=true]` and `[spring,spring-boot,useJakartaEe=true]`. Neither of those combinations is in the table I drafted. I'm not sure they are valid, so maybe I can't read straight. A sample project would be so much easier, and we could move the long, distracting conversation over to there.

**Comment by boardtc** [comment]
Thanks for sticking with me. I know if you didn't; no one would as this is just one of 2000+ bugs multiple pages down the list that no one ever visits unless they encounter this problem. I will try and see if I can create a sscce, you will be the only one looking at it I think. The [sscce](https://github.com/brunobottazzini/openapi-generator-json) for the [14270 bug](https://github.com/OpenAPITools/openapi-generator/issues/14270), which I am now getting was never commented on by anyone.

Apologies, the [config](https://github.com/OpenAPITools/openapi-generator/issues/13124#issuecomment-1387451058) is updated now. I had moved useJakartaEe to the first execution. But are you saying, Java, useSpringBoot3, useJakartaEe is not valid? But Spring Boot 3 uses Jakarta, that's the nub of the whole thing,

```
       <execution>
            <id>openapi-server</id>
            <phase>generate-sources</phase>
            <goals>
              <goal>generate</goal>
            </goals>
            <configuration>
              <inputSpec>${project.basedir}/src/main/resources/v1/foo.yaml</inputSpec>
              <generatorName>java</generatorName>
              <output>src/main/resources/v1</output>
              <configOptions>
                <useSpringBoot3>true</useSpringBoot3>
               <useJakartaEe>true</useJakartaEe>
                <interfaceOnly>true</interfaceOnly>
              </configOptions>
            </configuration>
          </execution>
          <execution>
            <id>openapi-client</id>
            <phase>generate-sources</phase>
            <goals>
              <goal>generate</goal>
            </goals>
            <configuration>
              <generatorName>spring</generatorName>
              <modelNamePrefix>foo</modelNamePrefix>
              <configOptions>
                <library>spring-boot</library>
              </configOptions>
              <apiPackage>foo</apiPackage>
              <modelPackage>foo.model</modelPackage>
              <groupId>${project.groupId}</groupId>
              <artifactId>${project.artifactId}</artifactId>
              <artifactVersion>${project.version}</artifactVersion>
              <environmentVariables>
                <enablePostProcessFile>true</enablePostProcessFile>
              </environmentVariables>
            </configuration>
          </execution>
```

**Comment by dsyer** [comment]
Here's what I see in that fragment:

| id | generatorName | library | options |
|----|----------------------|----------|-----------|
|openapi-server | java | | useSpringBoot3=true, useJakartaEe=true |
|openapi-client | spring | spring-boot | |

So the "server" execution generates a client, and it doesn't use a library so there won't be any Spring. And the "client" generates a server, and there is no `useSpringBoot3` so it won't use the Jakarta namespace. At least one of us is still confused.

**Comment by boardtc** [comment]
Many thanks for sticking with me. I guess I am confused too. With the `previous` version of the plugin all we are generating is:
```
[INFO] writing file fooApi.java
[INFO] writing file foo\generated-sources\openapi\pom.xml
[INFO] writing file foo\generated-sources\openapi\README.md
[INFO] writing file foo\generated-sources\openapi\src\java\main\com\bar\ApiUtil.java
[INFO] writing file foo\generated-sources\openapi\.openapi-generator-ignore
[INFO] writing file foo\generated-sources\openapi\.openapi-generator\VERSION
[INFO] writing file foo\generated-sources\openapi\.openapi-generator\FILES
```
This si what I want to reproduce with the new version.

So maybe I only need the client? Should I combine my 2 executions into 1?

**Comment by dsyer** [comment]
It's hard to say. Your reluctance to provide a sample is a problem. I could hazard a guess that since you generate `ApiUtil` you are generating a server, but that's only an educated guess.

**Comment by skagedal** [comment]
Came here from [this Stack Overflow-comment](https://stackoverflow.com/questions/74593513/is-there-a-way-to-configure-openapi-generator-to-use-jakarta-package-during-gene#comment131930906_74601642) and just wanted to say that `'useJakartaEe': 'true'` works beautifully for me with the `java` generator and the `native` library. Thanks for good work!

**Comment by christian-ertl** [comment]
The feign generator still creates a build.gradle (and build.sbt) with the wrong jakarta version (with javax package), the generated pom.xml is correct.

**Comment by wing328** [comment]
I wonder if you can file a PR to fix it 🙏

**Comment by wing328** [comment]
@douggish may I know if you can help contribute a PR to fix it?

a good starting point is to compare the following:

- https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/pom.mustache
- https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/build.gradle.mustache

**Comment by douggish** [comment]
> @douggish may I know if you can help contribute a PR to fix it?
> 
> a good starting point is to compare the following:
> 
> * https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/pom.mustache
> * https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/build.gradle.mustache

I would have to ask for permission from my company. What is the process for contributing? Is there a Contributor License Agreement that would need to be signed, etc.?

**Comment by wing328** [comment]
No need to sign CLA

https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-submit-a-pr-to-fix-bugs-or-make-enhancements is a good starting point.

**Comment by douggish** [comment]
Alright, I created a PR: https://github.com/OpenAPITools/openapi-generator/pull/20322

**Comment by wing328** [comment]
PR merged. Thanks for the contribution.

**Comment by ftaurino-reply** [comment]
Using `<generatorName>java</generatorName>` and `<library>microprofile</library>` + `<useJakartaEe>true</useJakartaEe>` generates `import javax`. Any update on this?

**Comment by z-aki** [comment]
> Using `<generatorName>java</generatorName>` and `<library>microprofile</library>` + `<useJakartaEe>true</useJakartaEe>` generates `import javax`. Any update on this?

`<additionalProperties>rootJavaEEPackage=jakarta</additionalProperties>`

**Comment by z-aki** [comment]
> Using `<generatorName>java</generatorName>` and `<library>microprofile</library>` + `<useJakartaEe>true</useJakartaEe>` generates `import javax`. Any update on this?

`<additionalProperties>rootJavaEEPackage=jakarta</additionalProperties>`

@ftaurino-reply

**Comment by railsharipov** [comment]
When `useJakartaEE="true"` and `library="okhttp-gson"` jakarta annotation library version is too old.
Potential workaround:
```
tasks.named('openApiGenerate').configure {
    doLast {
        def generatedBuildFile = file("build/generated-code/build.gradle")
        if (generatedBuildFile.exists()) {
            def content = generatedBuildFile.text
            content = content.replaceAll(
                /jakarta_annotation_version = "1\.[\d\.]*"/,
                'jakarta_annotation_version = "2.1.0"'
            )
            generatedBuildFile.text = content
        }
    }
}
```
